# Data Cleaning with Pandas

In this notebook we'll go through a few basic data cleaning steps that should be performed on all new datasets where necessary.

We'll go through the process with both the `orders` and `orderlines` datasets. You can then practice these skills by cleaning the `products` dataset yourself

In [1]:
import pandas as pd

In [2]:
# orders.csv
url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orders = pd.read_csv(path)

# orderlines.csv
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orderlines = pd.read_csv(path)

Before we begin, let's create a copy of the `orders` and `orderlines` DataFrames. This way we are sure any of our changes won't affect the original DataFrames

In [3]:
orders_df = orders.copy()

orderlines_df = orderlines.copy()

One of the best ways to begin data cleaning is by exploring using `.info()`. This will tell us:
* The shape of the DataFrame
* The names of the columns
* If there are any missing values
* The datatypes of the columns

By exploring the missing values and correcting any incorrect datatypes, we often come across inconsistencies in our data.

Beyond this, we should also have a **check for any duplicate rows**.

Let's first deal with the duplicates, as it's nice and easy, then we'll explore what `.info()` has to tell us.

## 1.&nbsp; Duplicates
We can check for duplicates using the pandas [.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) method.

We can then delete these rows, if we wish, using [.drop_duplicates()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

In [4]:
# orders_df
orders_df.duplicated().sum()

0

In [5]:
# orderlines_df
orderlines_df.duplicated().sum()

0

We have no duplicate rows in either DataFrame. Easy, there is no problem to solve. Normally though, if there were some duplicates, we'd drop the extra rows.

# 2.&nbsp; `.info()`

In [6]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


* `total_paid` has 5 missing values
* `created_date` should become datetime datatype

In [7]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


* `date` should be a datetime datatype
* `unit_price` should be a float datatype

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [8]:
print(f"5 missing values represents {((orders_df.total_paid.isna().sum() / orders_df.shape[0])*100).round(5)}% of the rows in our DataFrame")

5 missing values represents 0.0022% of the rows in our DataFrame


A quick way to find out a percentage here, if you don't need to print out a sentence for yourself/students/colleagues is `.value_count(normalize=True)`

In [9]:
orders_df.total_paid.isna().value_counts(normalize=True)

False    0.999978
True     0.000022
Name: total_paid, dtype: float64

As there is such a tiny amount of missing values, we will simply delete these rows, as we have enough data without them.

In [10]:
orders_df = orders_df.loc[~orders.total_paid.isna(), :]

Should you have a significant number of missing values in the future, you have a choice:
+ you can impute the values
+ you can take the values from other DataFrames, if they are present there
+ you can delete the values
+ or any number of other creative solutions

Please, always consider how much time you have on your project, and what impact your method of choice will have on your final assesment.

### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines_df`

## 4.&nbsp; Datatypes

### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype

In [11]:
orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])

<ipython-input-11-87c26cc5d8b0>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])


### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype
* `unit_price` should be a float datatype

#### 4.1.1.&nbsp; `date`

In [12]:
orderlines_df["date"] = pd.to_datetime(orderlines_df["date"])

#### 4.1.2.&nbsp;`unit_price`

In [13]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

ValueError: ignored

As you can see when we try to convert `unit_price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have 2 decimal points. Let's see if there are any other numbers like this.

In [14]:
orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()

False    257814
True      36169
Name: unit_price, dtype: int64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [15]:
two_dot_percentage = ((orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines_df.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage}% of the rows in our DataFrame")

The 2 dot problem represents 12.3% of the rows in our DataFrame


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and we might even end up losing a larger portion of our data than this too. For the moment we will delete the rows as we only have 2 weeks for this project and I'd like some quick, accurate results to show. If we have time at the end, we can come back and investigate this problem further, maybe there's a solution?

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), we need to remove the whole order and not just the affected row. If we only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

We therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [16]:
two_dot_order_ids_list = orderlines_df.loc[orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]
orderlines_df = orderlines_df.loc[~orderlines_df.id_order.isin(two_dot_order_ids_list)]

In [17]:
orderlines_df.shape[0]

216250

We still have 216250 rows in orderlines to work with. This should be more than enough for our evaluation.

Now that all of the 2 decimal point prices have been removed, let's try again to convert the column `unit_price` to the correct datatype.

In [18]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

It worked perfectly

# Challenge: Clean the `products` DataFrame
Now it's your turn. Use the lessons you learnt above and clean the products DataFrame. You don't have to copy exactly what we did. Think about the consequences of your actions, sometimes it is ok to delete rows, other times you may wish to come up with more creative solutions.

In [19]:
# products.csv
url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
products = pd.read_csv(path)

In [20]:
products_df = products.copy()

In [21]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


### Look for Duplicates

In [22]:
products_df.duplicated().sum()

8746

In [23]:
products_df.drop_duplicates(inplace=True)
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10573 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


### Look for Missing values


In [24]:
products_df.price.isna().value_counts(normalize=True)

False    0.995652
True     0.004348
Name: price, dtype: float64

In [25]:
products_df[products_df.price.isna()].head()

sku                                               name  \
34    TWS0019  Twelve South MagicWand support Apple Magic Tra...   
1900  AII0008           Aiino Case MacBook Air 11 '' Transparent   
2039  CEL0020  Celly Ambo Luxury Leather Case + iPhone 6 Case...   
2042  CEL0007  Celly Wallet Case with removable cover Black i...   
2043  CEL0012            Celly Silicone Hard Shell iPhone 6 Blue   

                                                   desc price promo_price  \
34    MagicWand for wireless keyboard and Magic Trac...   NaN     299.899   
1900      MacBook Air 11-inch casing with matte finish.   NaN       22.99   
2039  Cover and housing together with magnet for iPh...   NaN     399.905   
2042             Case Book for iPhone 6 card case type.   NaN     128.998   
2043                      Hard Shell Silicone iPhone 6.   NaN        4.99   

      in_stock      type  
34           0      8696  
1900         0  13835403  
2039         0  11865403  
2042         0  11865403  
2043         0  11865403

In [26]:
three_digits_price = products_df.price.str.match(r'.*\.\d\d\d').fillna(False)
products_df.loc[three_digits_price, 'price'] = pd.to_numeric(products_df[three_digits_price].price.str.replace('.','',regex=False)) / 10000

In [28]:
products_df['price'] = pd.to_numeric(products_df['price'])

In [29]:
products_df[three_digits_price]

sku                                               name  \
362    REP0043                      Speaker lower repair iPhone 4   
480    PIE0011                     Internal Battery for iPhone 3G   
515    SEN0061  Sennheiser EZX 80 Handsfree iPhone iPad and iP...   
518    SEV0026  Service installation RAM + HDD + SSD MacBook /...   
525    SEV0024      Service installation RAM + HDD + SSD Mac mini   
...        ...                                                ...   
19312  REP0424                       Input repair Headphones iPad   
19313  REP0421                     iPad charging connector repair   
19314  REP0416                           iPad front camera repair   
19315  REP0413                            repair rear camera iPad   
19316  REP0407                                iPad battery repair   

                                                    desc     price  \
362    Repair service including parts and labor for i...   49.9004   
480          Replacement AC Adapter for Apple iPhone 3G.    9.8978   
515    IPhone bluetooth headset with microphone iPad ...   64.9891   
518    RAM + HDD installation + SSD in your MacBook /...   59.9918   
525    installation RAM HDD + SSD + on your Mac mini ...   59.9918   
...                                                  ...       ...   
19312  Repair service including parts and labor for iPad  699.9003   
19313  Repair service including parts and labor for iPad  699.9003   
19314  Repair service including parts and labor for iPad  699.9003   
19315  Repair service including parts and labor for iPad  699.9003   
19316  Repair service including parts and labor for iPad  699.9003   

      promo_price  in_stock      type  
362       499.004         0  1,44E+11  
480        98.978         0  21485407  
515       649.891         0      5384  
518       599.918         0  20642062  
525       599.918         0  20642062  
...           ...       ...       ...  
19312       69.99         0  1,44E+11  
19313       69.99         0  1,44E+11  
19314       69.99         0  1,44E+11  
19315       69.99         0  1,44E+11  
19316       69.99         0  1,44E+11  

[542 rows x 7 columns]

In [64]:
three_decimals = products_df.promo_price.str.match(r'.*\.\d{3}').fillna(False)

In [74]:
products_df.loc[three_decimals, 'promo_price'] = pd.to_numeric(products_df.promo_price[three_decimals].str.replace('.', '', regex=False))/10000
products_df.promo_price = pd.to_numeric(products_df.promo_price)

In [76]:
(
    products_df[products_df.price.isna()]
    .join(orderlines_df.set_index('sku'), on='sku', how='inner')
    .loc[lambda x: x.unit_price <= x.promo_price * 1.01]
)

sku                                               name  \
34     TWS0019  Twelve South MagicWand support Apple Magic Tra...   
34     TWS0019  Twelve South MagicWand support Apple Magic Tra...   
34     TWS0019  Twelve South MagicWand support Apple Magic Tra...   
2042   CEL0007  Celly Wallet Case with removable cover Black i...   
2042   CEL0007  Celly Wallet Case with removable cover Black i...   
...        ...                                                ...   
12100  CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12100  CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12100  CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12100  CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12100  CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   

                                                    desc price  promo_price  \
34     MagicWand for wireless keyboard and Magic Trac...   NaN      29.9899   
34     MagicWand for wireless keyboard and Magic Trac...   NaN      29.9899   
34     MagicWand for wireless keyboard and Magic Trac...   NaN      29.9899   
2042              Case Book for iPhone 6 card case type.   NaN      12.8998   
2042              Case Book for iPhone 6 card case type.   NaN      12.8998   
...                                                  ...   ...          ...   
12100                Ultra Thin Case for iPhone 6 and 6s   NaN       9.9898   
12100                Ultra Thin Case for iPhone 6 and 6s   NaN       9.9898   
12100                Ultra Thin Case for iPhone 6 and 6s   NaN       9.9898   
12100                Ultra Thin Case for iPhone 6 and 6s   NaN       9.9898   
12100                Ultra Thin Case for iPhone 6 and 6s   NaN       9.9898   

       in_stock      type       id  id_order  product_id  product_quantity  \
34            0      8696  1132171    303577           0                 1   
34            0      8696  1139821    306780           0                 1   
34            0      8696  1170003    319335           0                 1   
2042          0  11865403  1221025    342650           0                 1   
2042          0  11865403  1248845    356290           0                 1   
...         ...       ...      ...       ...         ...               ...   
12100         0  11865403  1262134    362612           0                 1   
12100         0  11865403  1273660    368331           0                 1   
12100         0  11865403  1288654    375502           0                 1   
12100         0  11865403  1292416    377230           0                 1   
12100         0  11865403  1303458    382501           0                 1   

       unit_price                date  
34          29.99 2017-01-06 12:33:30  
34          29.99 2017-01-10 20:17:42  
34          29.99 2017-02-02 18:51:08  
2042        12.90 2017-04-04 01:20:20  
2042        12.90 2017-05-16 14:13:03  
...           ...                 ...  
12100        9.99 2017-06-05 23:49:26  
12100        9.99 2017-06-23 08:43:18  
12100        9.99 2017-07-10 17:51:42  
12100        9.99 2017-07-14 01:06:39  
12100        9.99 2017-07-26 19:06:38  

[72 rows x 13 columns]

### Check / Change Data types

In [ ]:
# your code here